### PR TITLE
WebSocket NEXT: automatically close connection when OIDC extension provides SecurityIdentity and token expires

### LINKED ISSUE #40536: WebSockets Next: close the connection if the security identity has expired
### Description

This enhancement request is a follow up to #40312 (PR: #40534).
The idea is that the identity time validity constraints should not be dropped once a user moves from the authenticated HTTP session to the WS channel, i.e, if the user is required to re-authenticate or have the session cookie refreshed when working with HTTP, for example, when the OIDC Keycloak session expires after 1 hour, then it should also be the case if the user upgrades to WS. 

### Implementation ideas

OIDC already sets the following `SecurityIdentity` attribute:

```
public static final String QUARKUS_IDENTITY_EXPIRE_TIME = "quarkus.identity.expire-time";
``` 

which we agreed earlier with @stuartwdouglas about. At the moment it is only OIDC which sets it but other security mechanisms like `Form`, `WebAuthn`, can also set it later.

It is in seconds, so it is checked during the callback if it is less than the current time in seconds and the connection is closed if it is true.
 
The related issue is where the user is returned to once the connection is closed for the user to repeat the original HTTP request for another upgrade.

Lets say I'm authenticating with Google and now I'm looking at the Quarkus LangChain4j demo page with a chat bot icon, I press on it, I'm in the WS session... Now if the authentication has expired, I'd like to be returned to this initial demo page where I can click on the ChatBot icon again.

Adding a Web Sockets Next configuration property which will let the user be redirected to the configured HTTP page would be nice - for example, if this original HTTP page is an OIDC secured initial demo page, then OIDC will refresh the session and the user will be able to get back to the WS channel

### PR DESCRIPTION
**Pull request by michalvavrik**
- Closes: #40536

### CODE REVIEW

**Comment by sberyozkin** [comment]
Thanks @michalvavrik. Can users have a chance to react to this closure somehow, have the error handler called etc ?

**Comment by michalvavrik** [comment]
> Thanks @michalvavrik. Can users have a chance to react to this closure somehow, have the error handler called etc ?

Connection is getting closed, you shouldn't be able to stop it. What else can be done?

If we only fired on error, there are 4 strategies (close, log, call, do nothing, custom OnError method) and only the first one is suitable. Please take this with big reservation because I am no expert but IMO: anything but closing is not thread-safe because we cannot ensure that the identity is updated before next on message is called and we don't have way to stop everything it re-authentication happens (like token refresh).

I have added assertion that `OnClose` is called, hope it helps.

**Comment by sberyozkin** [comment]
@michalvavrik I'm thinking what the user experience will be like. Lets say we authenticate with Google into a Quarkus endpoint which returns HTML page, with an option to start a chatbot. The user clicks on it and a web socket session starts, so I guess we we have the larger HTML page with text like `Hello, Alice` and then a smaller embedded WS window where a chat bot conversation is ongoing. 

Lets say the session has expired. If now Alice presses on anything in the HTML page like a link to another HTML page, Quarkus may auto-refresh the session may or redirect Alice to Google to re-authenticate.
But what if, after the session has expired, Alice types something in the WS channel, what will she see in the ChatBot window if the connection is closed ?

**Comment by mkouba** [comment]
> I have added assertion that `OnClose` is called, hope it helps.

+1 

> Lets say the session has expired. If now Alice presses on anything in the HTML page like a link to another HTML page, Quarkus may auto-refresh the session or redirect Alice to Google to re-authenticate.
But what if, after the session has expired, Alice types something in the WS channel, what will she see in the ChatBot window if the connection is closed ?

It's up to the client I would say. I.e. if the connection is closed then the client (e.g. some JS code) can decide to redirect the user.

**Comment by sberyozkin** [comment]
Just curious, ideally, the user should realize that the ChatBot error related to the closed connection means it is time to refresh the surrounding HTML page

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/SecuritySupport.java [hunk 1]
```diff
@@ -29,4 +42,25 @@ void start() {
...
+    void onClose() {
+        if (onClose != null) {
+            onClose.run();
+        }
+    }
+
+    private static Runnable closeConnectionWhenIdentityExpired(Vertx vertx, WebSocketConnectionImpl connection,
+            SecurityIdentity identity) {
+        if (identity.getAttribute("quarkus.identity.expire-time") instanceof Long expireInSeconds) {
+            long timerId = vertx.setTimer((expireInSeconds * 1000) - System.currentTimeMillis(),
```
Comment (mkouba) [inline extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/SecuritySupport.java:54]: Why do you subtract the `System.currentTimeMillis()`? I would just use `vertx.setTimer(TimeUnit.SECOND.toMillis(expireInSeconds))`... i.e. run the action after the given time :thinking:.

Comment (sberyozkin) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/SecuritySupport.java:54]: expireInSeconds is a number of seconds from epoch

Comment (michalvavrik) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/SecuritySupport.java:54]: > expireInSeconds is a number of seconds from epoch

yep, the time accepts delay so it needs to be diff

Comment (mkouba) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/SecuritySupport.java:54]: > expireInSeconds is a number of seconds from epoch

Ugh, really? So it's not really `expireInSecods` but more like `expireAt`...

**Review by mkouba — APPROVED** [review]
Loogs good!

**Comment by michalvavrik** [comment]
> @michalvavrik I'm thinking what the user experience will be like. Lets say we authenticate with Google into a Quarkus endpoint which returns HTML page, with an option to start a chatbot. The user clicks on it and a web socket session starts, I guess in this case we have the larger HTML page with text like `Hello, Alice` and then a smaller embedded WS window where a chat bot conversation is ongoing.
> 
> Lets say the session has expired. If now Alice presses on anything in the HTML page like a link to another HTML page, Quarkus may auto-refresh the session or redirect Alice to Google to re-authenticate. But what if, after the session has expired, Alice types something in the WS channel, what will she see in the ChatBot window if the connection is closed ?

I agree with Martin's comments - client has to handle this.

I have one idea, but it would need to be a separate issue and I don't know if it makes sense - for a code flow we know that token is going to expire and if refreshing token is possible asynchronously via WebClient POST request, we could do that when the expiration time is getting closer. However at one time right before we synchronize there would be old token (not yet expired) used by the identity and a new issued token.

**Comment by sberyozkin** [comment]
>It's up to the client I would say. I.e. if the connection is closed then the client (e.g. some JS code) can decide to redirect the user.

Sure, this is what I was curious about, as long as the script can intercept this error and decide what to do, it is all good

**Comment by sberyozkin** [comment]
Hi @michalvavrik 

>for a code flow we know that token is going to expire and if refreshing token is possible asynchronously via WebClient POST request, we could do that when the expiration time is getting closer. However at one time right before we synchronize there would be old token (not yet expired) used by the identity and a new issued token.

This is Ok I guess, the WS session will still use valid identity until it expires, even if outside of the WS channel a new token is available... Let's discuss it further when you'd like

**Review by sberyozkin — APPROVED** [review]

**Comment by sberyozkin** [comment]
Proposing a backport to 3.11 as a hardening fix, which ensures the WS session can't go forever if the identity has expired

**Comment by gsmet** [comment]
I tried to backport that one but we would need to also backport https://github.com/quarkusio/quarkus/pull/40655 for the tests to pass.

I will let @mkouba decide what's best: either backport both or none (or adjust the tests).

**Comment by mkouba** [comment]
> I tried to backport that one but we would need to also backport #40655 for the tests to pass.
> 
> I will let @mkouba decide what's best: either backport both or none (or adjust the tests).

I would backport both... if possible. Thanks.

**Comment by gsmet** [comment]
> I would backport both... if possible. Thanks.

@mkouba done with 3.11.2

**Pull request merged**